service/telemetry_test.go

```diff
@@ -67,3 +87,133 @@ func TestBuildTelAttrs(t *testing.T) {
 	assert.Equal(t, "b", telAttrs[semconv.AttributeServiceVersion])
 	assert.Equal(t, "c", telAttrs[semconv.AttributeServiceInstanceID])
 }
+
+func TestTelemetryInit(t *testing.T) {
+	type metricValue struct {
+		value  float64
+		labels map[string]string
+	}
+
+	for _, tc := range []struct {
+		name            string
+		useOtel         bool
+		expectedMetrics map[string]metricValue
+	}{
+		{
+			name:    "UseOpenCensusForInternalMetrics",
+			useOtel: false,
+			expectedMetrics: map[string]metricValue{
+				metricPrefix + ocPrefix + counterName: {
+					value: 13,
+					labels: map[string]string{
+						"service_name":        "otelcol",
+						"service_version":     "latest",
+						"service_instance_id": testInstanceID,
+					},
+				},
+			},
+		},
+		{
+			name:    "UseOpenTelemetryForInternalMetrics",
+			useOtel: true,
+			// TODO: add a test to verify that OTel is emitting a `target_info`
+			// 	info metric with resource attributes https://github.com/open-telemetry/opentelemetry-go/issues/3166
+			expectedMetrics: map[string]metricValue{
+				metricPrefix + ocPrefix + counterName: {
+					value: 13,
+					labels: map[string]string{
+						"service_name":        "otelcol",
+						"service_version":     "latest",
+						"service_instance_id": testInstanceID,
+					},
+				},
+				metricPrefix + otelPrefix + counterName: {
+					value:  13,
+					labels: map[string]string{},
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			registry := featuregate.NewRegistry()
+			obsreportconfig.RegisterInternalMetricFeatureGate(registry)
+			require.NoError(t, registry.Apply(map[string]bool{obsreportconfig.UseOtelForInternalMetricsfeatureGateID: tc.useOtel}))
+
+			tel := newColTelemetry(registry)
+			buildInfo := component.NewDefaultBuildInfo()
+			cfg := telemetry.Config{
+				Resource: map[string]*string{
+					semconv.AttributeServiceInstanceID: &testInstanceID,
+				},
+			}
+
+			err := tel.initOnce(buildInfo, zap.NewNop(), cfg)
+			require.NoError(t, err)
+			defer func() {
+				require.NoError(t, tel.shutdown())
+			}()
+
+			v := createTestMetrics(t, tel.mp)
+			defer func() {
+				view.Unregister(v)
+			}()
+
+			// Wait some time because OpenCensus can take sometime to set up and collect a measure.
+			time.Sleep(10 * time.Millisecond)
```

Comment:
I had to add this casual wait because sometimes we're not getting OpenCensus metrics. I ran this a few times with `-test.count 1000` and it would just randomly fail. Adding a millisecond wait was enough for me to stop seeing it, but just for safety I bumped it to 10 milliseconds.

I suspect that this does not affect the existing integration simply because they do more work before starting to test.

Reply:
I think if you call `view.RetrieveData(ocPrefix + counterName)` this will work

Reply:
That did the trick, thanks @bogdandrutu!